File: Sources/agm/WinidLocator.swift
import Foundation

/// Finds the `winid` script via environment variables, the installed prefix, standard locations,
/// or the repo-adjacent development layout.
enum WinidLocator {
    /// `Sources/agm` → package root → parent (`tools`) → `winid`.
    private static var repoAdjacentWinid: URL {
        let here = URL(fileURLWithPath: #filePath, isDirectory: false)
        let pkgRoot = here.deletingLastPathComponent().deletingLastPathComponent().deletingLastPathComponent()
        return pkgRoot.deletingLastPathComponent().appendingPathComponent("winid")
    }

    static func resolve(environment env: [String: String] = ProcessInfo.processInfo.environment) -> URL? {
        var paths: [String] = []

        // 1. Explicit env overrides
        if let v = trim(env["NOTIFY_MAILBOX_WINID"]) { paths.append(v) }
        if let v = trim(env["WINID_SCRIPT"]) { paths.append(v) }

        // 2. Installed prefix: AGM_PREFIX/bin/winid (default ~/.agm)
        let prefix = trim(env["AGM_PREFIX"]) ?? "\(NSHomeDirectory())/.agm"
        paths.append("\(prefix)/bin/winid")

        // 3. Standard locations
        paths.append("\(NSHomeDirectory())/.local/bin/winid")
        paths.append("/usr/local/bin/winid")

        // 4. Repo-adjacent fallback (development)
        paths.append(repoAdjacentWinid.path)

        for raw in paths {
            let path = (raw as NSString).expandingTildeInPath
            guard !path.isEmpty else { continue }
            let url = URL(fileURLWithPath: path)
            if FileManager.default.isExecutableFile(atPath: url.path) {
                return url
            }
        }
        return nil
    }

    private static func trim(_ s: String?) -> String? {
        guard let s else { return nil }
        let t = s.trimmingCharacters(in: .whitespacesAndNewlines)
        return t.isEmpty ? nil : t
    }
}

File: src-tauri/src/services/winid_locator.rs
use std::path::PathBuf;

pub fn resolve() -> Option<PathBuf> {
    let env = std::env::vars().collect::<std::collections::HashMap<_, _>>();

    // Check env overrides
    for key in &["NOTIFY_MAILBOX_WINID", "WINID_SCRIPT"] {
        if let Some(val) = env.get(*key) {
            let path = PathBuf::from(val.trim());
            if is_executable(&path) {
                return Some(path);
            }
        }
    }

    // Check AGM_PREFIX/bin/winid
    let prefix = env
        .get("AGM_PREFIX")
        .map(|s| PathBuf::from(s.trim()))
        .unwrap_or_else(|| {
            dirs::home_dir()
                .unwrap_or_else(|| PathBuf::from("."))
                .join(".agm")
        });
    let agm_winid = prefix.join("bin").join("winid");
    if is_executable(&agm_winid) {
        return Some(agm_winid);
    }

    // Standard paths
    let candidates = [
        dirs::home_dir().map(|h| h.join(".local/bin/winid")),
        Some(PathBuf::from("/usr/local/bin/winid")),
        // Repo-adjacent fallback
        dirs::home_dir().map(|h| h.join("embitious/tools/winid")),
    ];

    for candidate in candidates.iter().flatten() {
        if is_executable(candidate) {
            return Some(candidate.clone());
        }
    }

    None
}

fn is_executable(path: &PathBuf) -> bool {
    use std::os::unix::fs::PermissionsExt;
    path.exists()
        && std::fs::metadata(path)
            .map(|m| m.permissions().mode() & 0o111 != 0)
            .unwrap_or(false)
}
